### PR TITLE
[SYCL] Remove default error code value in exception

### PR DESCRIPTION
**Pull request by s-kanaev**
Signed-off-by: Sergey Kanaev <sergey.kanaev@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/detail/platform_impl.hpp [hunk 1]
```diff
@@ -73,7 +73,7 @@ class platform_impl {
   cl_platform_id get() const {
     if (is_host())
       throw invalid_object_error(
-          "This instance of platform is a host instance");
+          "This instance of platform is a host instance", PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/platform_impl.hpp]:
```suggestion
          "This instance of platform is a host instance", CL_INVALID_PLATFORM );
```

File: sycl/include/CL/sycl/detail/platform_impl.hpp [hunk 2]
```diff
@@ -88,7 +88,7 @@ class platform_impl {
   const RT::PiPlatform &getHandleRef() const {
     if (is_host())
       throw invalid_object_error(
-          "This instance of platform is a host instance");
+          "This instance of platform is a host instance", PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/platform_impl.hpp]:
```suggestion
          "This instance of platform is a host instance", CL_INVALID_PLATFORM );
```

File: sycl/include/CL/sycl/detail/program_impl.hpp [hunk 3]
```diff
@@ -296,7 +296,8 @@ class program_impl {
     for (const auto &Device : Devices) {
       if (!Device.get_info<param>()) {
         throw feature_not_supported(
-            "Online compilation is not supported by this device");
+            "Online compilation is not supported by this device",
+            PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/program_impl.hpp]:
```suggestion
            CL_COMPILER_NOT_AVAILABLE);
```

File: sycl/include/CL/sycl/detail/queue_impl.hpp [hunk 4]
```diff
@@ -116,7 +117,8 @@ class queue_impl {
       return pi::cast<cl_command_queue>(MCommandQueue);
     }
     throw invalid_object_error(
-        "This instance of queue doesn't support OpenCL interoperability");
+        "This instance of queue doesn't support OpenCL interoperability",
+        PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/queue_impl.hpp]:
```suggestion
        PI_INVALID_OPERATION);
```

File: sycl/include/CL/sycl/types.hpp [hunk 5]
```diff
@@ -243,11 +243,11 @@ detail::enable_if_t<is_float_to_int<T, R>::value, R> convertImpl(T Value) {
     int OldRoundingDirection = std::fegetround();
     int Err = std::fesetround(FE_TONEAREST);
     if (Err)
-      throw runtime_error("Unable to set rounding mode to FE_TONEAREST");
+      throw std::runtime_error("Unable to set rounding mode to FE_TONEAREST");
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/types.hpp]: Why have you changed the type to std one?

File: sycl/source/detail/event_impl.cpp [hunk 6]
```diff
@@ -31,7 +31,8 @@ cl_event event_impl::get() const {
     return pi::cast<cl_event>(MEvent);
   }
   throw invalid_object_error(
-      "This instance of event doesn't support OpenCL interoperability.");
+      "This instance of event doesn't support OpenCL interoperability.",
+      PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/source/detail/event_impl.cpp]:
```suggestion
      PI_INVALID_OPERATION);
```

File: sycl/source/detail/event_impl.cpp [hunk 7]
```diff
@@ -69,7 +70,7 @@ event_impl::event_impl(RT::PiEvent Event, const context &SyclContext)
   if (MContext->is_host()) {
     throw cl::sycl::invalid_parameter_error(
         "The syclContext must match the OpenCL context associated with the "
-        "clEvent.");
+        "clEvent.", PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/source/detail/event_impl.cpp]:
```suggestion
        "clEvent.", PI_INVALID_CONTEXT);
```

File: sycl/source/detail/event_impl.cpp [hunk 8]
```diff
@@ -128,7 +129,8 @@ event_impl::get_profiling_info<info::event_profiling::command_submit>() const {
         this->getHandleRef(), this->getPlugin());
   }
   if (!MHostProfilingInfo)
-    throw invalid_object_error("Profiling info is not available.");
+    throw invalid_object_error("Profiling info is not available.",
+                               PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/source/detail/event_impl.cpp]: Something like CL_PROFILING_INFO_NOT_AVAILABLE ?

File: sycl/source/detail/event_impl.cpp [hunk 9]
```diff
@@ -140,7 +142,8 @@ event_impl::get_profiling_info<info::event_profiling::command_start>() const {
...
```
Comment (romanovvlad) [inline sycl/source/detail/event_impl.cpp]: The same.

File: sycl/source/detail/memory_manager.cpp [hunk 10]
```diff
@@ -354,7 +354,8 @@ static void copyH2H(SYCLMemObjI *SYCLMemObj, char *SrcMem,
       (SrcOffset != id<3>{0, 0, 0} || DstOffset != id<3>{0, 0, 0} ||
        SrcSize != SrcAccessRange || DstSize != DstAccessRange)) {
     assert(!"Not supported configuration of memcpy requested");
-    throw runtime_error("Not supported configuration of memcpy requested");
+    throw runtime_error("Not supported configuration of memcpy requested",
+                        PI_INVALID_VALUE);
```
Comment (romanovvlad) [inline sycl/source/detail/memory_manager.cpp]: INVALID_OPERATION ?

File: sycl/source/detail/platform_impl.cpp [hunk 11]
```diff
@@ -101,13 +101,13 @@ static std::vector<DevDescT> getWhiteListDesc() {
     }
 
     if (':' != *str)
-      throw sycl::runtime_error("Malformed device white list");
+      throw sycl::runtime_error("Malformed device white list", PI_INVALID_VALUE);
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.cpp]: Isn't it perfect candidate for CL_OUT_OF_RESOURCES or CL_OUT_OF_HOST_MEMORY :) ?

File: sycl/source/detail/program_impl.cpp [hunk 12]
```diff
@@ -400,7 +405,8 @@ void program_impl::create_pi_program_with_kernel_name(
 template <>
 cl_uint program_impl::get_info<info::program::reference_count>() const {
   if (is_host()) {
-    throw invalid_object_error("This instance of program is a host instance");
+    throw invalid_object_error("This instance of program is a host instance",
+                               PI_INVALID_DEVICE);
```
Comment (romanovvlad) [inline sycl/source/detail/program_impl.cpp]:
```suggestion
                               PI_INVALID_OPERATION);
```

File: sycl/include/CL/sycl/detail/platform_impl.hpp (hunk 1 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/platform_impl.hpp]: @romanovvlad shouldn't `CL_INVALID_PLATFORM` have a PI version?

Comment (s-kanaev) [reply sycl/include/CL/sycl/detail/platform_impl.hpp]: @alexbatashev I've added this one to the mapping (enum)

File: sycl/include/CL/sycl/types.hpp (hunk 5 above)
Comment (s-kanaev) [reply sycl/include/CL/sycl/types.hpp]: This exception is result of standard function error.

File: sycl/source/detail/platform_impl.cpp (hunk 11 above)
Comment (s-kanaev) [reply sycl/source/detail/platform_impl.cpp]: I'm really happy that it is not...

File: sycl/source/detail/program_impl.cpp (hunk 12 above)
Comment (sergey-semenov) [reply sycl/source/detail/program_impl.cpp]: Whenever SYCL reports an `invalid_object_error`, I think it's more logically consistent to report a `PI_INVALID_DEVICE`/`CONTEXT`/`PROGRAM`/etc. where we can instead of `PI_INVALID_OPERATION` (unless we can get more descriptive with something like `CL_PROFILING_INFO_NOT_AVAILABLE`).

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 13]
```diff
@@ -279,7 +279,8 @@ class HostKernel : public HostKernelBase {
     for (int I = 0; I < Dims; ++I) {
       if (NDRDesc.LocalSize[I] == 0 ||
           NDRDesc.GlobalSize[I] % NDRDesc.LocalSize[I] != 0)
-        throw sycl::nd_range_error("Invalid local size for global size");
+        throw sycl::nd_range_error("Invalid local size for global size",
+                                   PI_INVALID_VALUE);
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/detail/cg.hpp]:
```suggestion
                                   PI_INVALID_WORK_GROUP_SIZE);
```

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 14]
```diff
@@ -320,7 +321,8 @@ class HostKernel : public HostKernelBase {
...
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/detail/cg.hpp]:
```suggestion
                                   PI_INVALID_WORK_GROUP_SIZE);
```

File: sycl/include/CL/sycl/types.hpp (hunk 5 above)
Comment (AlexeySachkov) [reply sycl/include/CL/sycl/types.hpp]: Yep, but is thrown by SYCL runtime, not by standard function. I would prefer SYCL exception here

File: sycl/include/CL/sycl/intel/sub_group_host.hpp [hunk 15]
```diff
@@ -23,138 +23,166 @@ struct sub_group {
...
-    throw runtime_error("Subgroups are not supported on host device. ");
+    throw runtime_error("Subgroups are not supported on host device. ",
+                        PI_INVALID_DEVICE);
   }
 
   template <typename T, access::address_space Space>
   void store(multi_ptr<T, Space> dst, T &x) const {
-    throw runtime_error("Subgroups are not supported on host device. ");
+    throw runtime_error("Subgroups are not supported on host device. ",
+                        PI_INVALID_DEVICE);
```
Comment (AlexeySachkov) [inline sycl/include/CL/sycl/intel/sub_group_host.hpp]: I guess it is more like `feature_not_supported("text", PI_INVALID_OPERATION)`, but I'm okay with this resolution for now

File: sycl/source/detail/context_impl.cpp [hunk 16]
```diff
@@ -98,7 +98,8 @@ cl_context context_impl::get() const {
     return pi::cast<cl_context>(MContext);
   }
   throw invalid_object_error(
-      "This instance of context doesn't support OpenCL interoperability.");
+      "This instance of context doesn't support OpenCL interoperability.",
+      PI_INVALID_DEVICE);
```
Comment (AlexeySachkov) [inline sycl/source/detail/context_impl.cpp]:
```suggestion
      PI_INVALID_CONTEXT);
```

File: sycl/source/detail/program_impl.cpp (hunk 12 above)
Comment (romanovvlad) [reply sycl/source/detail/program_impl.cpp]: My way to interpret is "I call a method, get an exception saying INVALID_OPERATION, go to spec and understand that I should not call such method of the host program."
While if I get INVALID_CONTEXT I can think that the context is broken for some reason, but in reality it's perfectly valid.

I don't know what is more logical for other people, so I'm OK with Sergey's suggestion.

Comment (sergey-semenov) [reply sycl/source/detail/program_impl.cpp]: Personally, I do agree with that line of thinking, it's just that SYCL spec chooses to interpret these cases as "invalid objects", so I think our reported error codes should follow suit.

**Comment by romanovvlad** [comment]
@s-kanaev  ping

File: sycl/include/CL/sycl/types.hpp (hunk 5 above)
Comment (s-kanaev) [reply sycl/include/CL/sycl/types.hpp]: OK.
I'm not really sure which `PI_*` error would fit here. Thus, I'll employ PI_ERROR_UNKNOWN.

File: sycl/source/detail/event_impl.cpp (hunk 6 above)
Comment (s-kanaev) [reply sycl/source/detail/event_impl.cpp]: Done

File: sycl/source/detail/event_impl.cpp (hunk 7 above)
Comment (s-kanaev) [reply sycl/source/detail/event_impl.cpp]: Done

File: sycl/source/detail/event_impl.cpp (hunk 8 above)
Comment (s-kanaev) [reply sycl/source/detail/event_impl.cpp]: Done. There is PI_PROFILING_INFO_NOT_AVAILABLE now.

File: sycl/source/detail/event_impl.cpp (hunk 9 above)
Comment (s-kanaev) [reply sycl/source/detail/event_impl.cpp]: Done

File: sycl/source/detail/memory_manager.cpp (hunk 10 above)
Comment (s-kanaev) [reply sycl/source/detail/memory_manager.cpp]: Done.

File: sycl/source/detail/program_impl.cpp (hunk 12 above)
Comment (s-kanaev) [reply sycl/source/detail/program_impl.cpp]: Done.

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 13 above)
Comment (s-kanaev) [reply sycl/include/CL/sycl/detail/cg.hpp]: Done

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 14 above)
Comment (s-kanaev) [reply sycl/include/CL/sycl/detail/cg.hpp]: Done

File: sycl/source/detail/context_impl.cpp (hunk 16 above)
Comment (s-kanaev) [reply sycl/source/detail/context_impl.cpp]: Done

File: sycl/include/CL/sycl/types.hpp (hunk 5 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/types.hpp]: Are you going to change the type?

Comment (romanovvlad) [reply sycl/include/CL/sycl/types.hpp]: @s-kanaev ping

Comment (s-kanaev) [reply sycl/include/CL/sycl/types.hpp]: Oh, sorry, I missed the comment.
Yes, the type is changed to `cl::sycl::runtime_error` with `PI_ERROR_UNKNOWN`

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 17]
```diff
@@ -264,7 +264,7 @@ Command *Scheduler::GraphBuilder::insertMemoryMove(MemObjRecord *Record,
     AllocaCmdSrc = (Record->MAllocaCommands.end() != It) ? *It : nullptr;
   }
   if (!AllocaCmdSrc)
-    throw runtime_error("Cannot find buffer allocation");
+    throw runtime_error("Cannot find buffer allocation", PI_INVALID_VALUE);
```
Comment (sergey-semenov) [inline sycl/source/detail/scheduler/graph_builder.cpp]: I don't think `PI_INVALID_VALUE` is appropriate here, but this and some other exceptions here are internal errors and should probably be assertions instead, but this is outside the scope of this patch, so I'm fine with this resolution for now.

File: sycl/source/detail/sycl_mem_obj_t.cpp [hunk 18]
```diff
@@ -29,17 +29,19 @@ SYCLMemObjT::SYCLMemObjT(cl_mem MemObject, const context &SyclContext,
   if (MInteropContext->is_host())
     throw cl::sycl::invalid_parameter_error(
         "Creation of interoperability memory object using host context is "
-        "not allowed");
+        "not allowed",
+        PI_INVALID_OPERATION);
```
Comment (sergey-semenov) [inline sycl/source/detail/sycl_mem_obj_t.cpp]:
```suggestion
        PI_INVALID_CONTEXT);
```

File: sycl/source/device_selector.cpp [hunk 19]
```diff
@@ -27,7 +27,8 @@ device device_selector::select_device() const {
   if (res != nullptr)
     return *res;
 
-  throw cl::sycl::runtime_error("No device of requested type available.");
+  throw cl::sycl::runtime_error("No device of requested type available.",
+                                PI_INVALID_OPERATION);
```
Comment (sergey-semenov) [inline sycl/source/device_selector.cpp]:
```suggestion
                                PI_DEVICE_NOT_FOUND);
```

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 20]
```diff
@@ -236,7 +236,8 @@ class SYCLMemObjT : public SYCLMemObjI {
     if (useHostPtr())
       throw invalid_parameter_error(
           "Buffer constructor from a pair of iterator values does not support "
-          "use_host_ptr property.");
+          "use_host_ptr property.",
+          PI_INVALID_OPERATION);
```
Comment (garimagu) [inline sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: Do you think PI_INVALID_VALUE is not apt for a invalid_parameter_error.
You have used that in the above file pretty consistently.

File: sycl/include/CL/sycl/handler.hpp [hunk 21]
```diff
@@ -296,10 +296,12 @@ class handler {
...
     if (is_host()) {
       throw invalid_object_error(
-          "This kernel invocation method cannot be used on the host");
+          "This kernel invocation method cannot be used on the host",
+          PI_INVALID_DEVICE);
     }
     if (SyclKernel.is_host()) {
-      throw invalid_object_error("Invalid kernel type, OpenCL expected");
+      throw invalid_object_error("Invalid kernel type, OpenCL expected",
+                                 PI_RESULT_INVALID_KERNEL_NAME);
```
[the remark this answers is not in the document]
Comment (s-kanaev) [reply sycl/include/CL/sycl/handler.hpp]: Oh, really. That's a typo.
Done.

File: sycl/source/detail/kernel_impl.hpp [hunk 22]
```diff
@@ -73,7 +73,8 @@ class kernel_impl {
   /// @return a valid cl_kernel instance
   cl_kernel get() const {
     if (is_host())
-      throw invalid_object_error("This instance of kernel is a host instance");
+      throw invalid_object_error("This instance of kernel is a host instance",
+                                 PI_INVALID_DEVICE);
```
[the remark this answers is not in the document]
Comment (s-kanaev) [reply sycl/source/detail/kernel_impl.hpp]: Done.

File: sycl/source/detail/program_impl.cpp [hunk 23]
```diff
@@ -46,14 +47,16 @@ program_impl::program_impl(
     Prg->throw_if_state_is_not(program_state::compiled);
     if (Prg->MContext != MContext) {
       throw invalid_object_error(
-          "Not all programs are associated with the same context");
+          "Not all programs are associated with the same context",
+          PI_INVALID_CONTEXT);
```
[the remark this answers is not in the document]
Comment (s-kanaev) [reply sycl/source/detail/program_impl.cpp]: Done.

File: sycl/source/detail/program_impl.cpp [hunk 24]
```diff
@@ -152,7 +155,8 @@ program_impl::~program_impl() {
 cl_program program_impl::get() const {
   throw_if_state_is(program_state::none);
   if (is_host()) {
-    throw invalid_object_error("This instance of program is a host instance");
+    throw invalid_object_error("This instance of program is a host instance",
+                               PI_INVALID_DEVICE);
```
[the remark this answers is not in the document]
Comment (s-kanaev) [reply sycl/source/detail/program_impl.cpp]: Done

File: sycl/source/detail/program_impl.cpp [hunk 25]
```diff
@@ -245,7 +249,8 @@ kernel program_impl::get_kernel(string_class KernelName,
   throw_if_state_is(program_state::none);
   if (is_host()) {
     if (IsCreatedFromSource)
-      throw invalid_object_error("This instance of program is a host instance");
+      throw invalid_object_error("This instance of program is a host instance",
+                                 PI_INVALID_DEVICE);
```
[the remark this answers is not in the document]
Comment (s-kanaev) [reply sycl/source/detail/program_impl.cpp]: Done.

File: sycl/source/detail/sycl_mem_obj_t.cpp (hunk 18 above)
Comment (s-kanaev) [reply sycl/source/detail/sycl_mem_obj_t.cpp]: Done

File: sycl/source/device_selector.cpp (hunk 19 above)
Comment (s-kanaev) [reply sycl/source/device_selector.cpp]: Done.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp (hunk 20 above)
Comment (s-kanaev) [reply sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: There is no requirement in spec on exception in this case.
Though, for sake of consistency I'd rather change the type of exception to a mere `cl::sycl::runtime_error`.
Done.
